### PR TITLE
[Transaction] Consume transaction message logic optimized

### PR DESCRIPTION
**Pull request by gaoran10**
### Motivation

The transaction message consuming logic needs to be optimized. One transaction commit marker in the topic partition consists of many Entries in the TransactionBuffer, so one transaction could be seen as one batch.

### Modifications

1. When sending messages to consumers use the `batchIndex` in the MessageIdData to present the startBatchIndex of the Entry.

2. Calculate the transaction messages number.

3. When the first reading one transaction init it's batchDeletedIndexes in the ManagedCursorImpl.

### Verifying this change

This change can be verified as follows:

- *org.apache.pulsar.broker.transaction.TransactionConsumeTest*

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (no)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java [hunk 1]
```diff
@@ -2857,5 +2857,14 @@ private int applyMaxSizeCap(int maxEntries, long maxSizeBytes) {
 
         return Math.min(maxEntriesBasedOnSize, maxEntries);
     }
+
+    public void internalInitBatchDeletedIndex(PositionImpl position, int totalNumMessageInBatch) {
+        if (!batchDeletedIndexes.containsKey(position)) {
+            BitSetRecyclable bitSetRecyclable = BitSetRecyclable.create();
+            bitSetRecyclable.set(0, totalNumMessageInBatch);
+            batchDeletedIndexes.put(position, bitSetRecyclable);
+        }
```
Comment (codelipenghui) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:2866]: batchDeletedIndexes.computeIfAbsent().

File: .github/workflows/ci-unit-broker-broker-gp1.yaml [hunk 2]
```diff
@@ -65,13 +65,13 @@ jobs:
...
-        run: ./build/retry.sh mvn -B -ntp test -pl pulsar-broker -Dinclude='**/TransactionProduceTest.java' -DtestForkCount=1 -DtestReuseFork=true
+        run: ./build/retry.sh mvn -B -ntp test -pl pulsar-broker -Dinclude='org/apache/pulsar/broker/transaction/**/*.java,org/apache/pulsar/client/transaction/**/*.java' -DfailIfNoTests=false -DtestForkCount=1 -DtestReuseFork=true
 
       - name: run unit tests for pulsar-broker
         if: steps.docs.outputs.changed_only == 'no'
-        run: ./build/retry.sh mvn -B -ntp test -pl pulsar-broker -Dinclude='org/apache/pulsar/broker/**/*.java' -Dexclude='org/apache/pulsar/broker/zookeeper/**/*.java,org/apache/pulsar/broker/loadbalance/**/*.java,org/apache/pulsar/broker/service/**/*.java,**/AdminApiOffloadTest.java,**/TransactionProduceTest.java'
+        run: ./build/retry.sh mvn -B -ntp test -pl pulsar-broker -Dinclude='org/apache/pulsar/broker/**/*.java' -Dexclude='org/apache/pulsar/broker/zookeeper/**/*.java,org/apache/pulsar/broker/loadbalance/**/*.java,org/apache/pulsar/broker/service/**/*.java,**/AdminApiOffloadTest.java,org/apache/pulsar/broker/transaction/**/*.java,org/apache/pulsar/client/transaction/**/*.java'
```
Comment (codelipenghui) [inline .github/workflows/ci-unit-broker-broker-gp1.yaml:74]: Shall we need this change?

**Review by codelipenghui — COMMENTED** [review]
Looks good, could you please add a test for the batch indexes of a transaction at the consumer side?

File: .github/workflows/ci-unit-broker-broker-gp1.yaml (hunk 2 above)
Comment (gaoran10) [reply .github/workflows/ci-unit-broker-broker-gp1.yaml:74]: I want to sperate transaction tests to make the transaction tests are executed in serial.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java (hunk 1 above)
Comment (gaoran10) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java]: I'll use this method instead.

**Review by codelipenghui — APPROVED** [review]

**Pull request merged**